Table of Contents

1. Import Libraies

2. Import Data

3. Wrangle the Data (Customer Dataframe)

4. Data Consistency Checks

5. Import ords_prods_merge dataframe

6. Check Data Types of Both Dataframes and Change if Applicable

7. Combine df_customers with df_ords_prods_merge on 'user_id' column and check merge

8. Export Dataframe

Question 3. Import Libraries and New Customer Data Set

1. Import Libraries

In [35]:
# Import libraries
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

2. Import Data

In [36]:
# Create path
path = '/Users/ianfleming/Desktop/05-2025 Instacart Basket Analysis'

In [37]:
# Import new customer data set as dataframe
df_customers = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'))

In [38]:
# Check import
df_customers

user_id First Name    Surnam  Gender           STATE  Age date_joined  \
0         26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1         33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2         65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3        125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4        130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
...         ...        ...       ...     ...             ...  ...         ...   
206204   168073       Lisa      Case  Female  North Carolina   44    4/1/2020   
206205    49635     Jeremy   Robbins    Male          Hawaii   62    4/1/2020   
206206   135902      Doris  Richmond  Female        Missouri   66    4/1/2020   
206207    81095       Rose   Rollins  Female      California   27    4/1/2020   
206208    80148    Cynthia     Noble  Female        New York   55    4/1/2020   

        n_dependants fam_status  income  
0                  3    married  165665  
1                  0     single   59285  
2                  2    married   99568  
3                  0     single   42049  
4                  1    married   40374  
...              ...        ...     ...  
206204             1    married  148828  
206205             3    married  168639  
206206             2    married   53374  
206207             1    married   99799  
206208             1    married   57095  

[206209 rows x 10 columns]

Question 4. Wrangle the Data 

3. Wrangle the Data

Rename columns for consistencies

In [39]:
# Rename columns
df_customers.rename(columns = {'First Name' : 'first_name', 'Surnam' : 'last_name', 'Gender' : 'gender', 'STATE' : 'state', 'Age' : 'age', 'n_dependants' : 'n_dependents'}, inplace = True)

In [40]:
# Check column names
df_customers.columns

Index(['user_id', 'first_name', 'last_name', 'gender', 'state', 'age',
       'date_joined', 'n_dependents', 'fam_status', 'income'],
      dtype='object')

4. Data Consistency Checks 

Checking Data Types

In [41]:
# Check data types
df_customers.dtypes

user_id          int64
first_name      object
last_name       object
gender          object
state           object
age              int64
date_joined     object
n_dependents     int64
fam_status      object
income           int64
dtype: object

In [42]:
# Check unique values in 'gender' column
df_customers['gender'].unique()

array(['Female', 'Male'], dtype=object)

In [43]:
# Check unique values in 'fam_status' column
df_customers['fam_status'].unique()

array(['married', 'single', 'living with parents and siblings',
       'divorced/widowed'], dtype=object)

In [44]:
# Check unique values in 'state' column
df_customers['state'].unique()

array(['Missouri', 'New Mexico', 'Idaho', 'Iowa', 'Maryland', 'Kentucky',
       'Montana', 'South Carolina', 'Texas', 'Virginia', 'Nevada',
       'Nebraska', 'Georgia', 'Wyoming', 'Colorado', 'North Dakota',
       'Wisconsin', 'Alaska', 'Vermont', 'Arkansas', 'Maine',
       'North Carolina', 'West Virginia', 'Indiana', 'Oregon', 'Florida',
       'California', 'Pennsylvania', 'Ohio', 'Connecticut', 'Arizona',
       'Louisiana', 'Washington', 'New York', 'Mississippi', 'Oklahoma',
       'Utah', 'New Hampshire', 'Hawaii', 'District of Columbia',
       'Alabama', 'Massachusetts', 'Rhode Island', 'Michigan',
       'New Jersey', 'Kansas', 'South Dakota', 'Minnesota', 'Illinois',
       'Tennessee', 'Delaware'], dtype=object)

In [45]:
# Check counts in 'gender' column
df_customers['gender'].value_counts(dropna=False)

gender
Male      104067
Female    102142
Name: count, dtype: int64

In [46]:
# Check counts in 'fam_status' column
df_customers['fam_status'].value_counts(dropna=False)

fam_status
married                             144906
single                               33962
divorced/widowed                     17640
living with parents and siblings      9701
Name: count, dtype: int64

In [48]:
# Change 'user_id' column to a string
df_customers['user_id'] = df_customers['user_id'].astype(str)


I changed 'user_id' column to a string because I do not need this to perform statistics on

In [49]:
# Check the data type change
df_customers['user_id'].dtype

dtype('O')

Checking for Missing Values

In [50]:
# Check for missing values
df_customers.isnull().sum()

user_id             0
first_name      11259
last_name           0
gender              0
state               0
age                 0
date_joined         0
n_dependents        0
fam_status          0
income              0
dtype: int64

The dataframe has 11,259 missing first names. 

In [51]:
# Investigate missing values
df_customers_nan = df_customers[df_customers['first_name'].isnull() == True]

In [52]:
df_customers_nan

user_id first_name last_name  gender         state  age date_joined  \
53       76659        NaN   Gilbert    Male      Colorado   26    1/1/2017   
73       13738        NaN     Frost  Female     Louisiana   39    1/1/2017   
82       89996        NaN    Dawson  Female        Oregon   52    1/1/2017   
99       96166        NaN   Oconnor    Male      Oklahoma   51    1/1/2017   
105      29778        NaN    Dawson  Female          Utah   63    1/1/2017   
...        ...        ...       ...     ...           ...  ...         ...   
206038  121317        NaN    Melton    Male  Pennsylvania   28   3/31/2020   
206044  200799        NaN  Copeland  Female        Hawaii   52    4/1/2020   
206090  167394        NaN     Frost  Female        Hawaii   61    4/1/2020   
206162  187532        NaN     Floyd  Female    California   39    4/1/2020   
206171  116898        NaN   Delgado  Female      Colorado   23    4/1/2020   

        n_dependents fam_status  income  
53                 2    married   41709  
73                 0     single   82518  
82                 3    married  117099  
99                 1    married  155673  
105                3    married  151819  
...              ...        ...     ...  
206038             3    married   87783  
206044             2    married  108488  
206090             1    married   45275  
206162             0     single   56325  
206171             2    married   59222  

[11259 rows x 10 columns]

I am going to leave the missing first names as NaN so no action is required.

Checking for Duplicates

In [53]:
# Check for duplicates
df_customers_dups = df_customers[df_customers.duplicated()]

In [54]:
# Check result
df_customers_dups

Empty DataFrame
Columns: [user_id, first_name, last_name, gender, state, age, date_joined, n_dependents, fam_status, income]
Index: []

There were no duplicates found

In [55]:
# Double check the key column for duplicates
df_customers['user_id'].duplicated().sum()

0

Checking for Outliers or Strange Values

In [56]:
df_customers.describe()

age   n_dependents         income
count  206209.000000  206209.000000  206209.000000
mean       49.501646       1.499823   94632.852548
std        18.480962       1.118433   42473.786988
min        18.000000       0.000000   25903.000000
25%        33.000000       0.000000   59874.000000
50%        49.000000       1.000000   93547.000000
75%        66.000000       3.000000  124244.000000
max        81.000000       3.000000  593901.000000

The max income is very high 593,901 and needs further investigation

In [59]:
# Check high incomes
df_customers[df_customers['income'] > 500000].sort_values(by='income', ascending=False)

user_id first_name   last_name  gender          state  age date_joined  \
21726    58596     Thomas       Cohen    Male     Washington   75    5/6/2017   
202655   98580        NaN   Patterson    Male     Washington   43   3/12/2020   
81583   103702    Kathryn  Mclaughlin  Female      Louisiana   55   4/14/2018   
111950  187680  Catherine      Church  Female        Wyoming   43   10/7/2018   
167037  189861     Pamela      Jordan  Female   Pennsylvania   47   8/21/2019   
...        ...        ...         ...     ...            ...  ...         ...   
80006   168516      Nancy   Gutierrez  Female         Hawaii   48    4/5/2018   
174110  114554     Carlos      Medina    Male       Delaware   55   9/30/2019   
143953  167103    Rebecca          Yu  Female       New York   45   4/10/2019   
91847   197521      Kathy    Mitchell  Female  West Virginia   81   6/12/2018   
47924    61454    Beverly    Townsend  Female      Wisconsin   47   10/3/2017   

        n_dependents fam_status  income  
21726              2    married  593901  
202655             2    married  592409  
81583              3    married  591089  
111950             2    married  590790  
167037             0     single  584097  
...              ...        ...     ...  
80006              3    married  504053  
174110             2    married  503088  
143953             1    married  502865  
91847              1    married  501962  
47924              1    married  500871  

[70 rows x 10 columns]

There are 70 rows of customers earning over 500,000. Because I am not sure if they are typos or actual income, I will use the median rather than the mean for income analysis. 

5. Import ords_prods_merge dataframe

In [61]:
# Import ords_prods_merge datarame
df_ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge.pkl'))

In [63]:
# Check import
df_ords_prods_merge

order_id  user_id  order_number  orders_day_of_week  \
0          2539329        1             1                   2   
1          2539329        1             1                   2   
2          2539329        1             1                   2   
3          2539329        1             1                   2   
4          2539329        1             1                   2   
...            ...      ...           ...                 ...   
32434484   2977660   206209            13                   1   
32434485   2977660   206209            13                   1   
32434486   2977660   206209            13                   1   
32434487   2977660   206209            13                   1   
32434488   2977660   206209            13                   1   

          order_hour_of_day  days_since_prior_order  first_order  product_id  \
0                         8                     NaN         True         196   
1                         8                     NaN         True       14084   
2                         8                     NaN         True       12427   
3                         8                     NaN         True       26088   
4                         8                     NaN         True       26405   
...                     ...                     ...          ...         ...   
32434484                 12                     7.0        False       14197   
32434485                 12                     7.0        False       38730   
32434486                 12                     7.0        False       31477   
32434487                 12                     7.0        False        6567   
32434488                 12                     7.0        False       22920   

          add_to_cart_order  reordered  ... match    price_range_loc  \
0                         1          0  ...  both  Mid–range product   
1                         2          0  ...  both  Mid–range product   
2                         3          0  ...  both  Low-range product   
3                         4          0  ...  both  Low-range product   
4                         5          0  ...  both  Low-range product   
...                     ...        ...  ...   ...                ...   
32434484                  5          1  ...  both  Mid–range product   
32434485                  6          0  ...  both  Mid–range product   
32434486                  7          0  ...  both  Low-range product   
32434487                  8          0  ...  both  Low-range product   
32434488                  9          0  ...  both  Low-range product   

            busiest_days  busiest_period_of_day max_order      loyalty_flag  \
0         Regularly busy         Average orders        10      New customer   
1         Regularly busy         Average orders        10      New customer   
2         Regularly busy         Average orders        10      New customer   
3         Regularly busy         Average orders        10      New customer   
4         Regularly busy         Average orders        10      New customer   
...                  ...                    ...       ...               ...   
32434484    Busiest days            Most orders        13  Regular customer   
32434485    Busiest days            Most orders        13  Regular customer   
32434486    Busiest days            Most orders        13  Regular customer   
32434487    Busiest days            Most orders        13  Regular customer   
32434488    Busiest days            Most orders        13  Regular customer   

         avg_price_per_user spending_flag  median_days_per_user  \
0                  6.367797   Low spender                  20.5   
1                  6.367797   Low spender                  20.5   
2                  6.367797   Low spender                  20.5   
3                  6.367797   Low spender                  20.5   
4                  6.367797   Low spender                  20.5   
...                     ...           ...           

Since I am going to combine the two dataframes using 'user_id' column, I will check the data type of this column in both. I previously changed 'user_id' to a string in df_customers and perhaps this was incorrect. Let's see.

6. Check Data Types of Both Dataframes and Change if Applicable

In [64]:
# Check data types of 'user:id' column in df_ords_prods_merge
df_ords_prods_merge['user_id'].dtype

dtype('int64')

In [65]:
# Check data types of 'uder_id' in df_customers
df_customers['user_id'].dtype

dtype('O')

Since I do not want to do any statistics on 'user_id', I want to keep it as a string so I will convert 'user_id' to string in the df_ords_prods_merge dataframe. 

In [66]:
# Change 'user_id' column to a string
df_ords_prods_merge['user_id'] = df_ords_prods_merge['user_id'].astype(str)

In [67]:
# Check change
df_ords_prods_merge['user_id'].dtype

dtype('O')

Question 6 . Combining Both DataFrames

7. Combine df_customers with df_ords_prods_merge on 'user_id' column and check merge

I am using a left join as i do not want to lose any orders in my dataframe.

In [68]:
# Combine dataframes
df_merged = df_ords_prods_merge.merge(df_customers, on='user_id', how='left', indicator = 'merge_flag')

In [69]:
# Check merge
df_merged['merge_flag'].value_counts()

merge_flag
both          32434489
left_only            0
right_only           0
Name: count, dtype: int64

In [70]:
# Check shape
df_merged.shape

(32434489, 34)

In [71]:
# Check columns
df_merged.head()

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329       1             1                   2                  8   
1   2539329       1             1                   2                  8   
2   2539329       1             1                   2                  8   
3   2539329       1             1                   2                  8   
4   2539329       1             1                   2                  8   

   days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                     NaN         True         196                  1   
1                     NaN         True       14084                  2   
2                     NaN         True       12427                  3   
3                     NaN         True       26088                  4   
4                     NaN         True       26405                  5   

   reordered  ... first_name  last_name  gender    state age date_joined  \
0          0  ...      Linda     Nguyen  Female  Alabama  31   2/17/2019   
1          0  ...      Linda     Nguyen  Female  Alabama  31   2/17/2019   
2          0  ...      Linda     Nguyen  Female  Alabama  31   2/17/2019   
3          0  ...      Linda     Nguyen  Female  Alabama  31   2/17/2019   
4          0  ...      Linda     Nguyen  Female  Alabama  31   2/17/2019   

  n_dependents fam_status  income merge_flag  
0            3    married   40423       both  
1            3    married   40423       both  
2            3    married   40423       both  
3            3    married   40423       both  
4            3    married   40423       both  

[5 rows x 34 columns]

In [72]:
# check column names
df_merged.columns

Index(['order_id', 'user_id', 'order_number', 'orders_day_of_week',
       'order_hour_of_day', 'days_since_prior_order', 'first_order',
       'product_id', 'add_to_cart_order', 'reordered', 'product_name',
       'aisle_id', 'department_id', 'prices', 'match', 'price_range_loc',
       'busiest_days', 'busiest_period_of_day', 'max_order', 'loyalty_flag',
       'avg_price_per_user', 'spending_flag', 'median_days_per_user',
       'order_frequency_flag', 'first_name', 'last_name', 'gender', 'state',
       'age', 'date_joined', 'n_dependents', 'fam_status', 'income',
       'merge_flag'],
      dtype='object')

In [73]:
# Drop the 'match' column from the previous merge
df_merged.drop(columns='match', inplace=True)

In [74]:
df_merged.shape

(32434489, 33)

In [75]:
# Check merge using an outer join
df_check = df_ords_prods_merge.merge(df_customers, on='user_id', how='outer', indicator='merge_flag')

In [76]:
df_check['merge_flag'].value_counts()

merge_flag
both          32434489
left_only            0
right_only           0
Name: count, dtype: int64

df_customers matched perfectly. i didn't lose any rows or add mismatched data. 

8. Export Dataframe

In [ ]:
# Export newly merged dataframe 'df_merged'

In [77]:
df_merged.to_pickle(os.path.join(path, '02 data', 'Prepared Data', 'ords_prods_cust_merged.pkl'))